// @nx/js@23.0.1/dist/src/utils/versions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.minSupportedTypescriptVersion = exports.typescriptVersion = exports.verdaccioVersion = exports.typesNodeVersion = exports.tsLibVersion = exports.swcNodeVersion = exports.swcHelpersVersion = exports.swcCoreVersion = exports.swcCliVersion = exports.prettierVersion = exports.esbuildVersion = exports.nxVersion = void 0;
const path_1 = require("path");
exports.nxVersion = require((0, path_1.join)('@nx/js', 'package.json')).version;
exports.esbuildVersion = '^0.27.0';
exports.prettierVersion = '~3.6.2';
exports.swcCliVersion = '~0.8.0';
exports.swcCoreVersion = '~1.15.5';
exports.swcHelpersVersion = '~0.5.18';
exports.swcNodeVersion = '~1.11.1';
exports.tsLibVersion = '^2.3.0';
exports.typesNodeVersion = '^22.0.0';
exports.verdaccioVersion = '^6.3.2';
// Typescript
exports.typescriptVersion = '~5.9.2';
/**
 * The minimum version is currently determined from the lowest version
 * that's supported by the lowest Angular supported version, e.g.
 * `npm view @angular/compiler-cli@18.0.0 peerDependencies.typescript`
 */
exports.minSupportedTypescriptVersion = '5.4.0';
